main/main.cpp

```diff
@@ -2090,23 +2090,39 @@ bool Main::iteration() {
 
 	iterating--;
 
-	if (fixed_fps != -1)
+	if (fixed_fps != -1) {
 		return exit;
-
-	if (OS::get_singleton()->is_in_low_processor_usage_mode() || !OS::get_singleton()->can_draw())
-		OS::get_singleton()->delay_usec(OS::get_singleton()->get_low_processor_usage_mode_sleep_usec()); //apply some delay to force idle time
-	else {
-		uint32_t frame_delay = Engine::get_singleton()->get_frame_delay();
-		if (frame_delay)
-			OS::get_singleton()->delay_usec(Engine::get_singleton()->get_frame_delay() * 1000);
 	}
 
-	int target_fps = Engine::get_singleton()->get_target_fps();
-	if (target_fps > 0 && !Engine::get_singleton()->is_editor_hint()) {
-		uint64_t time_step = 1000000L / target_fps;
+	const uint32_t frame_delay = Engine::get_singleton()->get_frame_delay();
+	if (frame_delay) {
+		// Add fixed frame delay to decrease CPU/GPU usage. This doesn't take
+		// the actual frame time into account.
+		// Due to the high fluctuation of the actual sleep duration, it's not recommended
+		// to use this as a FPS limiter.
+		OS::get_singleton()->delay_usec(frame_delay * 1000);
+	}
+
+	const int target_fps = Engine::get_singleton()->get_target_fps();
+	if ((target_fps > 0 && !Engine::get_singleton()->is_editor_hint()) ||
+			OS::get_singleton()->is_in_low_processor_usage_mode() ||
+			!OS::get_singleton()->can_draw()) {
```

Comment:
Isn't `!OS::get_singleton()->can_draw()` falling through the cracks of this refactoring with no `time_step`?

Reply:
I'd suggest refactoring for clarity (first indentation level is function scope):
```cpp
uint64_t dynamic_delay = 0;
if (OS::get_singleton()->is_in_low_processor_usage_mode() || !OS::get_singleton()->can_draw()) {
	dynamic_delay = OS::get_singleton()->get_low_processor_usage_mode_sleep_usec();
}
if (target_fps > 0 && !Engine::get_singleton()->is_editor_hint()) {
	dynamic_delay = MAX(dynamic_delay, (uint64_t)(1000000 / target_fps));
}

if (dynamic_delay > 0) {
	target_ticks += dynamic_delay;
	...
}
```